### ISSUE TITLE
Tests: Make ImportWarnings a failure

### ISSUE THREAD

**Issue by MSeifert04** [astropy/astropy#9224]
While running the tests I noticed an ImportWarning from NumPy (#9223). I think such a warning deserves being noticed by our CI. 

Maybe we should promote it to an Exception like we do with deprecations?

**Comment by pllim** [comment]
Doesn't hurt. :+1: from me.

cc @bsipocz

**Comment by bsipocz** [comment]
Yes, this is very much a deprecation, not sure why a new warning was used instead of it, I assume it has a story upstream if anyone cares to dig it up.

**Comment by MSeifert04** [comment]
The relevant commit was https://github.com/numpy/numpy/pull/10850 - however no mention why all `numpy.testing.utils` got a an importwarning and everything else a deprecation.

**Comment by bsipocz** [comment]
Thanks @MSeifert04 for digging the issue up. Based on the discussion there, it's an oversight, so we might not need to address this issue (though I suppose it cannot hurt to have the infrastructure to extend our machinery to be able to raise more types of warnings).
(and while we're there, I already run into some issues with `enable_deprecations_as_exceptions`, will clean up the experience for an issue)

**Comment by pllim** [comment]
Yeah, I think `enable_deprecations_as_exceptions` is due for some refactoring... 😬

**Comment by pllim** [comment]
See #8506

**Comment by bsipocz** [comment]
For the record adding ImportWarning to #8506 resulted an INTERNALERROR in `astropy/io/tests/test_registry.py`, but only for the tests that run parallel.


The full log is here:
```
 $ ~/munka/devel/astropy [:3f3c5ec90|…1⚑ 21] $ pt --parallel=4
running test
installing to temporary directory: /private/var/folders/dc/hsm7tqpx2d57n7vb3k1l81xw0000gq/T/astropy-test-c15qhgp0
warning: no files found matching '*' under directory 'scripts'
no previously-included directories found matching 'build'
warning: no files found matching '*.pyx' under directory 'astropy_helpers/astropy_helpers'
warning: no files found matching '*.h' under directory 'astropy_helpers/astropy_helpers'
warning: no files found matching '*.rst' under directory 'astropy_helpers/astropy_helpers'
no previously-included directories found matching 'astropy_helpers/build'
warning: no previously-included files matching '*.o' found anywhere in distribution
======================================================================================= test session starts ========================================================================================
platform darwin -- Python 3.7.4, pytest-5.1.2, py-1.8.0, pluggy-0.12.0

Running tests with Astropy version 4.0.dev25771.
Running tests in astropy docs.

Date: 2019-09-23T20:49:29

Platform: Darwin-17.7.0-x86_64-i386-64bit

Executable: /usr/local/opt/python/bin/python3.7

Full Python Version: 
3.7.4 (default, Jul  9 2019, 18:15:00) 
[Clang 10.0.0 (clang-1000.11.45.5)]

encodings: sys: utf-8, locale: UTF-8, filesystem: utf-8
byteorder: little
float info: dig: 15, mant_dig: 15

Numpy: 1.18.0.dev0+4d13d4c
Scipy: 1.3.1
Matplotlib: 3.1.1.post2162+g770d59d42
h5py: 2.8.0
Pandas: 0.24.2
Cython: 0.29.13
Scikit-image: 0.15.0
asdf: not available
astropy_helpers: 3.1rc1
Using Astropy options: remote_data: none.

rootdir: /private/var/folders/dc/hsm7tqpx2d57n7vb3k1l81xw0000gq/T/astropy-test-c15qhgp0/lib/python3.7/site-packages, inifile: setup.cfg
plugins: xdist-1.29.0, forked-1.0.2, arraydiff-0.3, cov-2.6.0, xonsh-0.8.3, dependency-0.4.0, remotedata-0.3.2, doctestplus-0.4.0, openfiles-0.4.0, requests-mock-1.6.0
gw0 [13733] / gw1 [13733] / gw2 [13733] / gw3 [13733]                                                                                                                                              
.................xxx...........x..x................................................................................................................x....x................................... [  1%]
..................................................x.x.........x............x.......xx........x........x........x.....................................................................x...... [  2%]
............................................................................................................................................................................................ [  4%]
............................................................................................................................................................................................ [  5%]
............................................................................................................................................................................................ [  6%]
............................................................................................................................................................................................. [  8%]
................................................s....s...........................................................................x..x............................x.......................... [  9%]
.........s.................s.......x........................................sssssss.sssss....sssssssssssss.ssssssss...sss...s.x.x..s..s.s................................................... [ 10%]
..................................s.ss.sss.ssss.s......s.s.s..............s........xxx................................................................s......s............s................. [ 12%]
.................xss.....s......................................................sss.ssssss.ssss.......ssss.sss......ss................s..xx..x....s.................ss...................... [ 13%]
..................................................................x..............................sss...........................................s..ss.............ss...........sss............ [ 15%]
x....x.............................................................ssss..ss........ss.s.s............s.................x......................................................s.......ssss.s [ 16%]
s.......sss...............................................sss..x...x................x...s.........................................................................s..........s....xxx..s.ss. [ 17%]
...................s......x........................s.........................................s.........................................s.ss.ss.ss.ssss.ss......s.sssssss....sss.........x... [ 19%]
.sxs..ss....s................................................................................sss............................................................................................ [ 20%]
...............................................................................................................................s............................s........................x...... [ 21%]
.s...............................s..................................s......sss.ss...s......................s.s.............................................................................. [ 23%]
.....................................................................................................................................................................s...................... [ 24%]
.................................................ss.s.s............s..............ss.s.s.................................................................................................... [ 26%]
.......................................x....................................................................................................s..........................s.........x.......... [ 27%]
............................................................................................................................................................................................ [ 28%]
............................................................s......F.........s...................................F.......F.....sF....F........F........................................F.... [ 30%]
......F........F..F......................................................................................................................................................................... [ 31%]
............................................................................................................................................................................................ [ 32%]
.............................F............................s.........F..F..F.FFF.............................................................................................................. [ 34%]
....................................................................................................................s.................s..................................................... [ 35%]
............................................................................................................................................................................................ [ 36%]
........................................................................................................................................................F...F......F.....F...............F.. [ 38%]
..F....F......F...F..F...F....F......................................................................................................................................F..................F... [ 39%]
.F.......F...F........F.....F.............................................F.F..F...............F..F...F..F.F...F............................................................................ [ 41%]
..........................................................................................s...s...sss..sssss.ss.......................s.............s.s..s..s............................... [ 42%]
.......................................s...s.....s.ss....s.....s............................................................................................................................ [ 43%]
............................................................................................................................................................................................ [ 45%]
............................................................................................................................................................................................ [ 46%]
..............................................................x............................................................................................................................. [ 47%]
......................................................................................................................................................................................s..... [ 49%]
.......................................................WARNING: failed to download https://maia.usno.navy.mil/ser7/finals2000A.all and https://toshi.nofs.navy.mil/ser7/finals2000A.all, using local IERS-B: <urlopen error An attempt was made to connect to the internet by a test that was not marked `remote_data`. The requested host was: maia.usno.navy.mil>;<urlopen error An attempt was made to connect to the internet by a test that was not marked `remote_data`. The requested host was: toshi.nofs.navy.mil> [astropy.utils.iers.iers]
...........................................................................x.................................................s....... [ 50%]
.......................................s.......ss................................ss.............................s........................................................................... [ 52%]
.........x..................................................................s....s..............................................................................................ss.......... [ 53%]
............................................................................................................................................................................................ [ 54%]
............................................................................................................................................................................................ [ 56%]
............................................................................................................................................................................................ [ 57%]
............................................................................................................................................................................................. [ 58%]
............................................................................................................................................................................................. [ 60%]
............................................................................................................................................................................................ [ 61%]
.....................................................................................................................................................................s....................... [ 63%]
............................................................................................................................................................................................ [ 64%]
............................................................................................................................................................................................ [ 65%]
.................................................................................ss...................................s.s.s.s.ss............................................................ [ 67%]
............................................................................................................................................................................................ [ 68%]
.......x.......x......x.......x......................x.......................x.............................................................................................................. [ 69%]
............x..............x......................x.........................................x.........................................................................................x..... [ 71%]
............................................................................................................................................................................................ [ 72%]
............................................................................................................................................................................................ [ 73%]
.............................................................................................................................................................................................. [ 75%]
............................................................................................................................................................................................ [ 76%]
............................................................................................................................................................................................ [ 78%]
.......................................................................................................................................................................x.................... [ 79%]
............................................................................................................................................................................................ [ 80%]
.................................................................................ss.......................................................................................................... [ 82%]
.....................................................................................................x...................................................................................... [ 83%]
.................................................................................x.........................................................x...............................................x. [ 84%]
..........s.s.............................................................s...ss............................................................................................................ [ 86%]
............................................................................................................................................................................................ [ 87%]
............................................................................................................................................................................................ [ 89%]
...........................................................................................................................................s................................................ [ 90%]
....................................................................................................................s....................................................................... [ 91%]
..........sss..................................................ss..sssss...........s.....s.s................................................................................................ [ 93%]
......................................ssss...ssss.sssssssssssssssssssssss.................ssss....F.........s.s............................ss............................................... [ 94%]
.....................................F.INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/_pytest/main.py", line 191, in wrap_session
INTERNALERROR>     session.exitstatus = doit(config, session) or 0
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/_pytest/main.py", line 235, in _main
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/hooks.py", line 289, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/manager.py", line 87, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/manager.py", line 81, in <lambda>
INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/callers.py", line 208, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/callers.py", line 80, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/callers.py", line 187, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "<remote exec>", line 68, in pytest_runtestloop
INTERNALERROR>   File "<remote exec>", line 85, in run_one_test
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/hooks.py", line 289, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/manager.py", line 87, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/manager.py", line 81, in <lambda>
INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/callers.py", line 208, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/callers.py", line 80, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/callers.py", line 187, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/_pytest/runner.py", line 72, in pytest_runtest_protocol
INTERNALERROR>     runtestprotocol(item, nextitem=nextitem)
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/_pytest/runner.py", line 87, in runtestprotocol
INTERNALERROR>     reports.append(call_and_report(item, "call", log))
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/_pytest/runner.py", line 171, in call_and_report
INTERNALERROR>     hook.pytest_runtest_logreport(report=report)
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/hooks.py", line 289, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/manager.py", line 87, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/manager.py", line 81, in <lambda>
INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/callers.py", line 208, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/callers.py", line 80, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/callers.py", line 187, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "<remote exec>", line 109, in pytest_runtest_logreport
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/hooks.py", line 289, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/manager.py", line 87, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/manager.py", line 81, in <lambda>
INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/callers.py", line 208, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/callers.py", line 80, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/callers.py", line 187, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/_pytest/reports.py", line 328, in pytest_report_to_serializable
INTERNALERROR>     data = report._to_json()
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/_pytest/reports.py", line 164, in _to_json
INTERNALERROR>     return _report_to_json(self)
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/_pytest/reports.py", line 394, in _report_to_json
INTERNALERROR>     d["longrepr"] = serialize_longrepr(report)
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/_pytest/reports.py", line 381, in serialize_longrepr
INTERNALERROR>     serialize_repr_crash(repr_crash),
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/_pytest/reports.py", line 367, in serialize_repr_crash
INTERNALERROR>     return reprcrash.__dict__.copy()
INTERNALERROR> AttributeError: 'NoneType' object has no attribute '__dict__'
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/_pytest/main.py", line 191, in wrap_session
INTERNALERROR>     session.exitstatus = doit(config, session) or 0
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/_pytest/main.py", line 235, in _main
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/hooks.py", line 289, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/manager.py", line 87, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/manager.py", line 81, in <lambda>
INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/callers.py", line 208, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/callers.py", line 80, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/pluggy/callers.py", line 187, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/xdist/dsession.py", line 115, in pytest_runtestloop
INTERNALERROR>     self.loop_once()
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/xdist/dsession.py", line 138, in loop_once
INTERNALERROR>     call(**kwargs)
INTERNALERROR>   File "/usr/local/lib/python3.7/site-packages/xdist/dsession.py", line 180, in worker_workerfinished
INTERNALERROR>     assert not crashitem, (crashitem, node)
INTERNALERROR> AssertionError: ('astropy/io/tests/test_registry.py::test_get_reader_invalid', <WorkerController gw3>)
INTERNALERROR> assert not 'astropy/io/tests/test_registry.py::test_get_reader_invalid'

======================================================== 47 failed, 12605 passed, 318 skipped, 65 xfailed, 1 warnings in 103.31s (0:01:43) =========================================================
$
```

**Comment by bsipocz** [comment]
@taldcroft - have you seen similar things regarding the registry?

**Comment by pllim** [comment]
Addressed in #8506.
